etcd3/stateful/watch.py

```diff
@@ -240,6 +240,24 @@ def onEvent(self, filter_or_cb, cb=None):
             raise TypeError('callback should be a callable')
         self.callbacks.append((self.get_filter(filter), filter, cb))
 
+    @check_param(at_least_one_of=['filter', 'cb'])
+    def unEvent(self, filter=None, cb=None):
+        """
+        remove a callback or filter event that's been previously added via onEvent()
+        If both parameters are given they are ANDd together; to OR the, make two calls.
+
+        :type filter: callable or regex string or EventType
+        :param filter: the callable filter or regex string or EventType the event to be removed was registerd with
+        :param cb: the callback funtion the event to be removed was registerd with
+        """
+        for i in reversed(range(len(self.callbacks))):
```

Comment:
I did a test and seems it won't raise an error when you change a list in other thread.

But you can not use for in range(length) here. Cause if you delete a element in other thread, length shorten, `self.callbacks[i]` could cause an "out of range" error

Reply:
that's why it's `reversed()`

Reply:
![image](https://user-images.githubusercontent.com/3241438/58587691-9d736e00-8290-11e9-98aa-f98a9ca126b2.png)
```python
from threading import Thread
import time
import sys

l = list(range(10))

def trd():
    while True:
        for i in reversed(range(len(l))):
            print('l[%d]: %d' % (i, l[i]))
            time.sleep(0.1)

t = Thread(target=trd)
t.setDaemon(True)
t.start()

def mod():
    while True:
        # l.append(1)
        del l[0]
        time.sleep(0.1)

t = Thread(target=mod)
t.setDaemon(True)
t.start()

time.sleep(5)
print('done')
sys.exit(0)
```

Reply:
There are two threads: the 'control' thread and the watcher thread:
  * The watcher thread just repeatedly does `for c in callbacks: ...`
  * The control thread is what can change the list, which is fine: https://repl.it/repls/TintedUntidyDesktopenvironment

Reply:
`for i in list` is fine, but iterate by indicies would cause error

Reply:
So now that it's protected by a lock, we just iterate over indices from largest to smallest so that a `del callbacks[i]` removes the one we mean it to.